### PR TITLE
Fixes #82: Add Jenkins App

### LINKED ISSUE #82: [Feature Request] Add Jenkins App
@alexellis want to have Jenkins as app in Arkade

## Expected Behaviour

Jenkins should be installed to the cluster and should be able to run jobs. In addition it would be nice, if Jenkins could spawn containers as worker nodes.

## Possible Solution

* Jenkins should be install via the https://github.com/helm/charts/tree/master/stable/jenkins helm chart. 
* Jenkins should run in its own namespace
* There should be proper resource limits for doing a maven build
* Jenkins should be reachable via LoadBalancer and with admin credentials
* The Kubernetes integration should be able to deploy applications and manifests to any namespace of the cluster

### PR DESCRIPTION
**Pull request by aidun**
Fixes #82: Add Jenkins App
## Description

- Uses Jenkins Helm Chart
- Helm 3 only
- no ARM support
- Only default settings
## Motivation and Context

No Jenkins app where present. @alexellis want it :-)

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

```
k3d create
./arkade install jenkins
kubectl get all
kubectl --namespace default port-forward svc/jenkins 8080:8080 &
printf $(kubectl get secret --namespace default jenkins -o jsonpath="{.data.jenkins-admin-user}" | base64 --decode);echo
printf $(kubectl get secret --namespace default jenkins -o jsonpath="{.data.jenkins-admin-password}" | base64 --decode);echo

# Login
echo http://127.0.0.1:8080
```
```
k3d create
./arkade install jenkins
kubectl get all
kubectl get pvc
```

k3d on MacOS with docker 4 desktop

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [x] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

**Comment by aidun** [comment]
Could not test it on ARM, because my PI´s aren't ready. On docker hub are no arm images, so I decided to the installation.

File: cmd/apps/jenkins_app.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		return nil
+	}
+
+	return jenkins
+}
+
+var JenkinsInfoMsg = `# Forward the Jenkins port to your machine
+kubectl --namespace default port-forward svc/jenkins 8080:8080 &
+
+# Get the admin-user and admin-password
```
Comment (alexellis) [inline cmd/apps/jenkins_app.go:128]: Any way you can break up this to work across 80 characters? We should try to honour a classic display width.

File: cmd/apps/jenkins_app.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		}
+
+		fmt.Println(jenkinsInstallMsg)
+		return nil
+	}
+
+	return jenkins
+}
+
+var JenkinsInfoMsg = `# Forward the Jenkins port to your machine
```
Comment (alexellis) [inline cmd/apps/jenkins_app.go:125]: We could change to:

`# To view the UI on http://127.0.0.1:8080/`

**Comment by alexellis** [comment]
Hi @aidun I'm struggling to get this to work on KinD

```
kind delete cluster && kind create cluster && go build && ./arkade install jenkins

LAST SEEN   TYPE      REASON                    OBJECT                          MESSAGE
2m55s       Warning   FailedScheduling          pod/jenkins-78c5b85c4d-nkv85    0/1 nodes are available: 1 node(s) had taints that the pod didn't tolerate.
2m25s       Normal    Scheduled                 pod/jenkins-78c5b85c4d-nkv85    Successfully assigned default/jenkins-78c5b85c4d-nkv85 to kind-control-plane
2m25s       Normal    Pulling                   pod/jenkins-78c5b85c4d-nkv85    Pulling image "jenkins/jenkins:lts"
2m21s       Normal    TaintManagerEviction      pod/jenkins-78c5b85c4d-nkv85    Cancelling deletion of Pod default/jenkins-78c5b85c4d-nkv85
2m55s       Normal    SuccessfulCreate          replicaset/jenkins-78c5b85c4d   Created pod: jenkins-78c5b85c4d-nkv85
2m55s       Normal    ScalingReplicaSet         deployment/jenkins              Scaled up replica set jenkins-78c5b85c4d to 1
3m16s       Normal    Starting                  node/kind-control-plane         Starting kubelet.
3m15s       Normal    NodeAllocatableEnforced   node/kind-control-plane         Updated Node Allocatable limit across pods
3m15s       Normal    NodeHasSufficientMemory   node/kind-control-plane         Node kind-control-plane status is now: NodeHasSufficientMemory
3m15s       Normal    NodeHasNoDiskPressure     node/kind-control-plane         Node kind-control-plane status is now: NodeHasNoDiskPressure
3m15s       Normal    NodeHasSufficientPID      node/kind-control-plane         Node kind-control-plane status is now: NodeHasSufficientPID
3m1s        Normal    RegisteredNode            node/kind-control-plane         Node kind-control-plane event: Registered Node kind-control-plane in Controller
3m          Normal    Starting                  node/kind-control-plane         Starting kube-proxy.
2m25s       Normal    NodeReady                 node/kind-control-plane         Node kind-control-plane status is now: NodeReady

```

**Comment by aidun** [comment]
🤔 I experienced some strange behavior as mentioned above in my first tests, but thought it was a bad connection issue. The last three tests with and without persistence worked fine. I don’t think it is an arkade issue, because I got this issues also with a helm install.

**Comment by aidun** [comment]
Is it the Init-Container which has the problems or the Jenkins container itself?

**Comment by alexellis** [comment]
```
kubectl get pod -w
NAME                       READY   STATUS     RESTARTS   AGE
jenkins-78c5b85c4d-rv2c2   0/1     Init:0/1   0          3m18s
jenkins-78c5b85c4d-rv2c2   0/1     Init:0/1   0          4m44s
jenkins-78c5b85c4d-rv2c2   0/1     PodInitializing   0          5m42s
jenkins-78c5b85c4d-rv2c2   0/1     Running           0          5m44s
jenkins-78c5b85c4d-rv2c2   1/1     Running           0          6m50s
```

I've been trying different things then eventually despite the errors it worked after 6 minutes. How long does it take for you now?

**Comment by aidun** [comment]
2-3 Minutes. But also to long. Maybe an issue with Ressource limits. I will check it tomorrow

________________________________
Von: Alex Ellis <notifications@github.com>
Gesendet: Monday, April 13, 2020 9:38:22 PM
An: alexellis/arkade <arkade@noreply.github.com>
Cc: Markus Hartmann <markush1986@gmail.com>; Mention <mention@noreply.github.com>
Betreff: Re: [alexellis/arkade] Fixes #82: Add Jenkins App (#90)


kubectl get pod -w
NAME                       READY   STATUS     RESTARTS   AGE
jenkins-78c5b85c4d-rv2c2   0/1     Init:0/1   0          3m18s
jenkins-78c5b85c4d-rv2c2   0/1     Init:0/1   0          4m44s
jenkins-78c5b85c4d-rv2c2   0/1     PodInitializing   0          5m42s
jenkins-78c5b85c4d-rv2c2   0/1     Running           0          5m44s
jenkins-78c5b85c4d-rv2c2   1/1     Running           0          6m50s


I've been trying different things then eventually despite the errors it worked after 6 minutes. How long does it take for you now?

**Comment by alexellis** [comment]
Thank you :+1:

**Comment by aidun** [comment]
The init-container have some strange starting times. From 30 seconds to 4 minutes. I have to look for the image...

**Pull request merged**